Annotate more of the binder

333fred opened the pull request.

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:
```diff
@@ -450,14 +450,14 @@ private BoundExpression CreateTupleLiteralConversion(SyntaxNode syntax, BoundTup
 
                     foreach (var argument in tupleSyntax.Arguments)
                     {
                         locationBuilder.Add(argument.NameColon?.Name.Location);
```
This is what is forcing all the `ImmutableArray<Location?>` annotation changes.

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_AnonymousTypes.cs:
```diff
@@ -203,13 +205,14 @@ private bool IsAnonymousTypesAllowed()
             {
                 if (expression.HasExpressionType())
                 {
+                    RoslynDebug.Assert(expressionType is object);
```
Consider `!` here given that `HasExpressionType` has already asserted this.

333fred replied:
[https://github.com/dotnet/roslyn/issues/40992](https://github.com/dotnet/roslyn/issues/40992). It would require a lot of `!`, so I'm going to keep this as is.

---
In reply to: [367115054](https://github.com/dotnet/roslyn/pull/40946#discussion_r367115054) [](ancestors = 367115054)

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -30,12 +31,12 @@ internal static void BindAttributeTypes(ImmutableArray<Binder> binders, Immutabl
             Debug.Assert(attributesToBind.Any());
             Debug.Assert((object)ownerSymbol != null);
```
Why is this even a parameter to the method?

333fred replied:
Honestly? No idea, but I think that refactoring is out of scope for a nullable annotation pr.

---
In reply to: [367115354](https://github.com/dotnet/roslyn/pull/40946#discussion_r367115354) [](ancestors = 367115354)

jaredpar replied:
I agree it's out of scope. I just noticed and couldn't hold back my comment.

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -102,7 +103,9 @@ internal CSharpAttributeData GetAttribute(AttributeSyntax node, NamedTypeSymbol
         internal BoundAttribute BindAttribute(AttributeSyntax node, NamedTypeSymbol attributeType, DiagnosticBag diagnostics)
         {
-            return this.GetBinder(node).BindAttributeCore(node, attributeType, diagnostics);
+            var binder = this.GetBinder(node);
```
Are we at the point that we should consider adding a `GetRequiredBinder` method?

333fred replied:
If not we soon will be. Will add.

---
In reply to: [367115570](https://github.com/dotnet/roslyn/pull/40946#discussion_r367115570) [](ancestors = 367115570)

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -111,15 +114,15 @@ private Binder SkipSemanticModelBinder()
         private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol attributeType, DiagnosticBag diagnostics)
         {
-            Debug.Assert(this.SkipSemanticModelBinder() == this.GetBinder(node).SkipSemanticModelBinder());
+            Debug.Assert(this.SkipSemanticModelBinder() == this.GetBinder(node)!.SkipSemanticModelBinder());
```
This is another place where `GetRequiredBinder` would help readability a bit.

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -1026,7 +1032,7 @@ private TypedConstant VisitExpression(BoundExpression node, DiagnosticBag diagno
             {
                 // Validate Statement 1) of the spec comment above.
 
-                var typedConstantKind = node.Type.GetAttributeParameterTypedConstantKind(_binder.Compilation);
```
Given `node.Type` is external to the method consider using `RoslynDebug.Assert` for this.

333fred commented:
@dotnet/roslyn-compiler for a second review.

jcouv commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3167,7 +3167,7 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
                 locationOpt: null, // no location for the type declaration
                 elementTypesWithAnnotations: typesBuilder.ToImmutableAndFree(),
-                elementLocations: elementLocations,
+                elementLocations: elementLocations!,
```
>elementLocations [](start = 34, length = 16)

Should the type of `elementLocations` be adjusted in public API then, so we avoid the suppression?

333fred replied:
I don't think so. If you're creating this from a public context, you should pass real locations as more of a matter of principle.

---
In reply to: [367179639](https://github.com/dotnet/roslyn/pull/40946#discussion_r367179639) [](ancestors = 367179639)

jcouv replied:
Just to clarify: When do we need to pass `null` locations internal?
Maybe we can mark as non-nullable across the board...

---
In reply to: [370338279](https://github.com/dotnet/roslyn/pull/40946#discussion_r370338279) [](ancestors = 370338279,367179639)

jcouv commented on src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs:
```diff
@@ -102,7 +102,7 @@ static NamedTypeSymbol getTupleUnderlyingType(ImmutableArray<TypeWithAnnotations
             ImmutableArray<string?> elementNames = default,
             ImmutableArray<bool> errorPositions = default,
-            ImmutableArray<Location> elementLocations = default,
+            ImmutableArray<Location?> elementLocations = default,
```
>ImmutableArray<Location?> elementLocations [](start = 12, length = 42)

what prompted that change? 
Do we need a test for `Compilation.CreateTupleTypeSymbol` passing null locations? (I couldn't find one)

333fred replied:
https://github.com/dotnet/roslyn/pull/40946#discussion_r366123920

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -332,13 +333,13 @@ protected bool IsAttributeConditionallyOmitted(NamedTypeSymbol attributeType, Sy
                         // TODO: use fully qualified identifier name for boundNamedArgumentsSet
-                        string argumentName = argument.NameEquals.Name.Identifier.ValueText;
+                        string argumentName = argument.NameEquals.Name.Identifier.ValueText!;
                         if (boundNamedArgumentsBuilder == null)
```
>boundNamedArgumentsBuilder == null [](start = 28, length = 34)

consider changing this test to check `boundNamedArgumentsSet` instead, since the two locals are always initialized together (or not), then the suppression can be removed.

333fred replied:
Our approach for nullable has been to not change any semantic behavior, just to annotate what we are currently doing. I'm going to keep this going forward.

---
In reply to: [367181664](https://github.com/dotnet/roslyn/pull/40946#discussion_r367181664) [](ancestors = 367181664)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -396,9 +397,9 @@ private BoundExpression BindNamedAttributeArgument(AttributeArgumentSyntax named
             // TODO: should we create an entry even if there are binding errors?
             var fieldSymbol = namedArgumentNameSymbol as FieldSymbol;
-            IdentifierNameSyntax nameSyntax = namedArgument.NameEquals.Name;
+            IdentifierNameSyntax nameSyntax = namedArgument.NameEquals!.Name;
```
>namedArgument.NameEquals [](start = 46, length = 24)

consider using an assertion instead

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:
```diff
@@ -425,10 +426,10 @@ private BoundExpression BindNamedAttributeArgument(AttributeArgumentSyntax named
         private Symbol BindNamedAttributeArgumentName(AttributeArgumentSyntax namedArgument, NamedTypeSymbol attributeType, DiagnosticBag diagnostics, out bool wasError, out LookupResultKind resultKind)
         {
-            var identifierName = namedArgument.NameEquals.Name;
+            var identifierName = namedArgument.NameEquals!.Name;
```
>NameEquals [](start = 47, length = 10)

consider using an assertion here too

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:
```diff
@@ -49,20 +50,20 @@ internal partial class Binder
                 var name = clause.Name.Identifier.ValueText;
                 int ordinal;
-                if (names.TryGetValue(name, out ordinal))
+                if (names.TryGetValue(name!, out ordinal))
```
>name [](start = 38, length = 4)

consider asserting instead

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:
```diff
@@ -98,7 +99,7 @@ internal partial class Binder
             TypeParameterConstraintClause.AdjustConstraintTypes(containingSymbol, typeParameters, results, ref isValueTypeOverride);
 
-            RemoveInvalidConstraints(typeParameters, results, syntaxNodes, diagnostics);
+            RemoveInvalidConstraints(typeParameters, results!, syntaxNodes, diagnostics);
```
>results [](start = 53, length = 7)

why this suppression?

333fred replied:
`RemoveInvalidConstraints` takes a non-null arraybuilder. We started with some nulls in the array, but just went through the whole thing to remove it.

---
In reply to: [367182789](https://github.com/dotnet/roslyn/pull/40946#discussion_r367182789) [](ancestors = 367182789)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:
```diff
@@ -74,7 +76,7 @@ internal partial class Binder
                 // We need to preserve any conversion that changes the type (even identity conversions, like object->dynamic),
                 // or that was explicitly written in code (so that GetSemanticInfo can find the syntax in the bound tree).
-                if (!isCast && source.Type.Equals(destination, TypeCompareKind.IgnoreNullableModifiersForReferenceTypes))
+                if (!isCast && source.Type!.Equals(destination, TypeCompareKind.IgnoreNullableModifiersForReferenceTypes))
```
>Type [](start = 38, length = 4)

how do we know `source.Type` isn't null from surrounding code?

333fred replied:
> Note: a few lines below we do a null-test on source.Type

`BindToNaturalType` _should_ take care of that, but I'll add an assert.

---
In reply to: [367183378](https://github.com/dotnet/roslyn/pull/40946#discussion_r367183378) [](ancestors = 367183378)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:
```diff
@@ -700,7 +700,7 @@ private bool MemberGroupFinalValidationAccessibilityChecks(BoundExpression recei
             else if (WasImplicitReceiver(receiverOpt))
             {
-                if (InFieldInitializer && !ContainingType.IsScriptClass || InConstructorInitializer || InAttributeArgument)
+                if (InFieldInitializer && !ContainingType!.IsScriptClass || InConstructorInitializer || InAttributeArgument)
```
>ContainingType [](start = 43, length = 14)

Is the suppression okay because `InFieldInitializer` is true at this point?

333fred replied:
Correct.

---
In reply to: [367183871](https://github.com/dotnet/roslyn/pull/40946#discussion_r367183871) [](ancestors = 367183871)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:
```diff
@@ -1067,14 +1067,14 @@ internal bool MethodGroupIsCompatibleWithDelegate(BoundExpression receiverOpt, b
             Debug.Assert(!sourceValue.IsBad);
 
             SpecialType destinationType;
             if ((object)destination != null && destination.IsEnumType())
```
>destination [](start = 24, length = 11)

Should `destination` be declared as nullable?

333fred replied:
The only callers pass non-null.

---
In reply to: [367184280](https://github.com/dotnet/roslyn/pull/40946#discussion_r367184280) [](ancestors = 367184280)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:
```diff
@@ -1093,7 +1093,7 @@ internal bool MethodGroupIsCompatibleWithDelegate(BoundExpression receiverOpt, b
                 {
                     // NOTE: Dev10 puts a suffix, "M", on the constant value.
-                    Error(diagnostics, ErrorCode.ERR_ConstOutOfRange, syntax, sourceValue.Value + "M", destination);
+                    Error(diagnostics, ErrorCode.ERR_ConstOutOfRange, syntax, sourceValue.Value + "M", destination!);
```
>destination [](start = 103, length = 11)

Can't `destination` be null here? Seems dangerous

333fred replied:
The null check above is actually superfluous.

---
In reply to: [367184489](https://github.com/dotnet/roslyn/pull/40946#discussion_r367184489) [](ancestors = 367184489)

jcouv replied:
Sorry, I'd missed your reply. Consider removing the superfluous null check so that we don't need those suppressions.
Suppressions are typically for when the compiler is not able to analyze properly, but here it's rather that the code is inconsistent (checks for null in one place, but would blow up with null in another).

---
In reply to: [370290111](https://github.com/dotnet/roslyn/pull/40946#discussion_r370290111) [](ancestors = 370290111,367184489)

333fred replied:
I'm not going to change the semantics of the code in this PR.

---
In reply to: [375590101](https://github.com/dotnet/roslyn/pull/40946#discussion_r375590101) [](ancestors = 375590101,370290111,367184489)

jcouv commented:
>             CSharpAttributeData[] attributesBuilder,

Should this hold nullable elements instead?

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:56 in 69941f7. [](commit_id = 69941f757708eed5a20e8620b40e4a70e3ceaf0d, deletion_comment = False)

jcouv commented:
>                 ConstantValue constantValue = node.ConstantValue;

Should this be nullable? we test for null below

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:1043 in 69941f7. [](commit_id = 69941f757708eed5a20e8620b40e4a70e3ceaf0d, deletion_comment = False)

333fred commented:
@jcouv addressed feedback. I haven't merged from master in a while, so I'll wait for your feedback and if you sign off, rebase and fix any new changes.

---
In reply to: [343620361](https://github.com/dotnet/roslyn/pull/40946#issuecomment-343620361) [](ancestors = 343620361)

333fred commented:
Verified this still builds locally.
